libs/@local/codegen/src/typescript.rs

```diff
@@ -0,0 +1,284 @@
+use oxc::{
+    allocator::Allocator,
+    ast::{AstBuilder, ast},
+    codegen::Codegen,
+    span::SPAN,
+};
+
+use crate::{
+    TypeCollection,
+    definitions::{Enum, EnumTagging, EnumVariant, Fields, Primitive, Type, TypeDefinition},
+};
+
+#[derive(Default)]
+#[non_exhaustive]
+pub struct TypeScriptGeneratorSettings {
+    allocator: Allocator,
+}
+
+#[expect(dead_code)]
+pub struct TypeScriptGenerator<'a> {
+    settings: &'a TypeScriptGeneratorSettings,
+    collection: TypeCollection,
+    ast: AstBuilder<'a>,
+}
+
+impl<'a> TypeScriptGenerator<'a> {
+    pub const fn new(
+        settings: &'a TypeScriptGeneratorSettings,
+        collection: TypeCollection,
+    ) -> Self {
+        Self {
+            settings,
+            collection,
+            ast: AstBuilder {
+                allocator: &settings.allocator,
+            },
+        }
+    }
+
+    #[must_use]
+    pub fn generate(&self, name: &str) -> String {
+        Codegen::new()
+            .build(
+                &self.ast.program(
+                    SPAN,
+                    ast::SourceType::d_ts(),
+                    "test",
+                    self.ast.vec(),
+                    None,
+                    self.ast.vec(),
+                    self.ast.vec1(
+                        self.visit_type_definition(name, &self.collection.types[name].1)
+                            .into(),
+                    ),
+                ),
+            )
+            .code
+    }
+
+    fn visit_type_definition(
```

Comment:
Do you think it'd make sense (in the future) to move this into a dedicated visitor, a lá https://github.com/hashintel/hash/blob/c041062cda2d075110b434c1db8801c4d6092b0f/libs/%40local/hashql/ast/src/visit.rs#L126 instead?

Reply:
That's over-engineering. It does not server any purpose to move it into a trait and add all kinds of different return types just to create a trait. It would create a trait just to have. trait which would be implemented exactly once. We can always add one later.
The goal is to get it working and improve over time, not come up with a pattern which is future proof long-term. We currently don't have any concrete plans to add more visitor implementations in the near future. A lot more things need to be considered before.